middleware/middleware.go

```diff
@@ -0,0 +1,48 @@
+package middleware
+
+import (
+	"net/http"
+)
+
+// Allowed provides a list of methods for which the handler should be executed
+type Allowed struct {
+	Methods []string
+	Handler func(w http.ResponseWriter, req *http.Request)
+}
+
+// isMethodAllowed determines if a method is allowed or not
+func (a *Allowed) isMethodAllowed(method string) bool {
+	for _, s := range a.Methods {
+		if method == s {
+			return true
+		}
+	}
+	return false
+}
+
+// HealthcheckFilter creates a map for healthcheck endpoints whitelisting, to be provided to the Whitelist generic function.
+// Usage example: 	middlewareChain := alice.New(middleware.Whitelist(middleware.HealthcheckFilter(hc.Handler)))
+func HealthcheckFilter(hcHandler func(w http.ResponseWriter, req *http.Request)) map[string]Allowed {
+	return map[string]Allowed{
+		"/health": Allowed{
+			Methods: []string{http.MethodGet},
+			Handler: hcHandler,
+		},
+	}
+}
+
+// Whitelist creates a middleware that executes whitelisted endpoints
+// The provided whitelist is keyed by path, and contains the handler to use and the methods for which the whitelist applies
+func Whitelist(whitelist map[string]Allowed) func(h http.Handler) http.Handler {
```

Comment:
I think nearly all this logic should sit with `go-ns/server` package; like so:

```
package server

import (
	"net/http"
)

// Allowed provides a list of methods for which the handler should be executed
type Allowed struct {
	Methods []string
	Handler func(w http.ResponseWriter, req *http.Request)
}

// isMethodAllowed determines if a method is allowed or not
func (a *Allowed) isMethodAllowed(method string) bool {
	for _, s := range a.Methods {
		if method == s {
			return true
		}
	}
	return false
}

// CreateWhitelist creates a map for allowed endpoints, to be provided to the Whitelist generic function
func CreateWhitelist(handler func(w http.ResponseWriter, req *http.Request), methods []string, path string) map[string]Allowed {
	return map[string]Allowed{
		path: Allowed{
			Methods: methods,
			Handler: Handler,
		},
	}
}

// Whitelist creates a middleware that executes whitelisted endpoints
// The provided whitelist is keyed by path, and contains the handler to use and the methods for which the whitelist applies
func Whitelist(whitelist map[string]Allowed) func(h http.Handler) http.Handler {
	return func(nextHandler http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
			if wh, ok := whitelist[req.URL.Path]; ok && wh.isMethodAllowed(req.Method) {
				wh.Handler(w, req)
				return
			}
			nextHandler.ServeHTTP(w, req)
		})
	}
}
```

And then here (dp-healthcheck) would be:

```
package middleware
import (
    "github.com/ONSDigital/go-ns/server"
)

// CreateWhitelist creates a map for healthcheck endpoints whitelisting, to be provided to the Whitelist generic function
func CreateWhitelist(hcHandler func(w http.ResponseWriter, req *http.Request)) map[string]server.Allowed {
	return server.CreateWhitelist(hcHandler, []string{http.MethodGet}, "/health")
}
```

Reply:
The reasoning is that the whitelist functionality may apply to other handlers used in the applications and actually has nothing to do with health checks even though we have used this for the first time for health checks?

We can still create a healthcheck whitelist as shown above.

Reply:
I can understand that separation, in principle, but wanted to let `go-ns` remain deprecated (rather than put more code in there).

Is now the time to create `dp-httpd` from `go-ns/server` ?  (better name suggestions welcome)

Reply:
dp-server? 

Also Im happy once the above function exists as then we can ship the rest of the code when we need to without needing to update the application code as the `CreateWhitelist` func will still exist here just now pointing to methods in a new/old library

Reply:
Assuming all other functions are not exported so that they cannot be called outside of dp-healthcheck........ or use the internal folder?

Reply:
`dp-server` may be better (since it'll be used as `server.Foo()` rather than the ambiguous (special?) `httpd.Foo()`

Not sure what functions you want private?